De-duplicate and improve definition of core::ffi::c_char

arichardson opened the pull request:
Instead of having a list of unsigned char targets for each OS, follow the logic Clang uses and instead set the value based on architecture with a special case for Darwin and Windows operating systems. This makes it easier to support new operating systems targeting Arm/AArch64 without having to modify this config statement for each new OS. The new list does not quite match Clang since I noticed a few bugs in the Clang implementation (https://github.com/llvm/llvm-project/issues/115957).

Fixes https://github.com/rust-lang/rust/issues/129945
Closes https://github.com/rust-lang/rust/pull/131319

arichardson committed:
De-duplicate and improve definition of core::ffi::c_char

Instead of having a list of unsigned char targets for each OS, follow the
logic Clang uses and instead set the value based on architecture with
a special case for Darwin and Windows operating systems. This makes it
easier to support new operating systems targeting Arm/AArch64 without
having to modify this config statement for each new OS. The new list does
not quite match Clang since I noticed a few bugs in the Clang
implementation (https://github.com/llvm/llvm-project/issues/115957).

Fixes: https://github.com/rust-lang/rust/issues/129945

arichardson commented:
r? @maurer @joshtriplett

maurer commented on library/core/src/ffi/mod.rs:
```diff
@@ -92,56 +92,46 @@ pub type c_ssize_t = isize;
-                    target_arch = "riscv32",
-                    target_arch = "csky"
-                )
+        // See isSignedCharDefault() in clang/lib/Driver/ToolChains/Clang.cpp
```
The comment above is no longer correct - at the end of this `if` block, you put `i8`

maurer commented on library/core/src/ffi/mod.rs:
```diff
@@ -92,56 +92,46 @@ pub type c_ssize_t = isize;
-                any(target_arch = "aarch64", target_arch = "riscv64")
+        ))] {
+            pub type c_char = i8;
+        } else if #[cfg(all(
```
We should probably only have one `if` and one `else`, one for `u8`, one for `i8`, having two `i8` branches should at worst be done by `any(`ing these two expressions together. We definitely shouldn't have 4 branches, 3 i8 and 1 u8.

arichardson replied:
I did it this way to avoid duplicating the 5 darwin OS combinations but as you say it might be cleaner to do it that way. Let me try that

arichardson replied:
I simplified this by relying on Apple operating systems other than macos only supporting Arm, so it's not too bad.

arichardson replied:
Thanks this is so much simpler!

maurer reviewed (changes requested):
Some OSes are no longer conditioned on, for example, os=`nto` arch=`aarch64` or os=`l4re`, arch=`x86_64` are missing.

I didn't manually verify everything, but at least `l4re` and `x86_64` just got flipped from `u8`, which it appears to have been manually set to, to `i8`.

Can we get a summary of what changed in terms of supported target triples? I would expect disagreements with `clang` to be rare, but this patch looks like it accidentally changes obscure targets at a minimum. It might also be clearer what changed if we didn't have the 4-branch if statement for 2 results.

maurer commented:
I'm pretty sure this still changes `l4re-x86_64` from unsigned to signed.

arichardson commented:
> I'm pretty sure this still changes `l4re-x86_64` from unsigned to signed.

I'm trying to figure out if that value is intentional or not. It dates back to https://github.com/rust-lang/rust/commit/2cf0a4ad4690c587d755dc6e5383779dddb9b178, but I can't see anything in L4RE that changes the default - maybe this was actually indented to set it for aarch64? CC: @TobiasSchaffner

arichardson commented:
Looks like l4re should be setting this for all architectures, I spotted `-funsigned-char` in https://github.com/kernkonzept/mk/blob/926afa93e32e64dbdb33cf9ae724924ee1fb16e0/tool/kconfig/Makefile#L550, so that suggests it's the case for all targets. Will update to handle that.

joshtriplett commented:
r? libs-api

tgross35 commented:
See also an open issue about this https://github.com/rust-lang/rust/issues/129945 (cc @taiki-e)

arichardson commented:
I have validated the new c_char definitions against my build of clang that has fixes for MSP430, Xtensa (although printing the builtin defines fails with `error: unknown target triple 'xtensa-unknown-none-elf'`), and CSKY (which fails with `version 'abiv2' in target triple 'csky-unknown-linux-gnuabiv2' is invalid`).

Unlike https://github.com/rust-lang/rust/pull/131319 this PR does not have links to all the ABI docs, but it does have a correct definition for l4re (which clang does not support and treats as unknown OS).

taiki-e commented:
As for L4Re:
I think L4Re code you are linking is config when building kernel. IIUC, userland uses signed for x86_64 and unsigned for aarch64 by default (i.e., same as ELF ABI's default).

```
$ ./toolchain-l4re-x86_64-gcc-14/bin/x86_64-l4re-gcc -E -dM -x c /dev/null | grep -F __CHAR_
#define __CHAR_BIT__ 8

$ ./toolchain-l4re-arm64-gcc-14/bin/aarch64-l4re-gcc -E -dM -x c /dev/null | grep -F __CHAR_
#define __CHAR_BIT__ 8
#define __CHAR_UNSIGNED__ 1
```

({x86_64,aarch64}-l4re-gcc are from https://l4re.org/download/snapshots/toolchain/.)

---

As for Xtensa:
Section 2.17.1 "Data Types and Alignment" of Xtensa LX Microprocessor Overview handbook <https://loboris.eu/ESP32/Xtensa_lx%20Overview%20handbook.pdf> says "`char` type is unsigned by default".


However, it appears that Xtensa has multiple ABIs. (Section 1.4 "Calling convention" of Overview of Xtensa ISA <https://dl.espressif.com/github_assets/espressif/xtensa-isa-doc/releases/download/latest/Xtensa.pdf>)

arichardson committed:
Remove l4re from the unsigned char operating system list

As noted in https://github.com/rust-lang/rust/pull/132975#issuecomment-2484645240,
the default for userland apps is to follow the architecture defaults, the
-funsigned-char flag only applies to kernel builds.

arichardson commented:
Rebased and fixed the l4re issue.

> As for L4Re: I think [L4Re code you are linking](https://github.com/rust-lang/rust/pull/132975#issuecomment-2474547410) is config when building kernel. IIUC, userland uses signed for x86_64 and unsigned for aarch64 by default (i.e., same as ELF ABI's default).
> 
> ```
> $ ./toolchain-l4re-x86_64-gcc-14/bin/x86_64-l4re-gcc -E -dM -x c /dev/null | grep -F __CHAR_
> #define __CHAR_BIT__ 8
> 
> $ ./toolchain-l4re-arm64-gcc-14/bin/aarch64-l4re-gcc -E -dM -x c /dev/null | grep -F __CHAR_
> #define __CHAR_BIT__ 8
> #define __CHAR_UNSIGNED__ 1
> ```
> 
> ({x86_64,aarch64}-l4re-gcc are from https://l4re.org/download/snapshots/toolchain/.)
> 

Thanks for checking this, I've updated the change to remove l4re from the list. I've done this as a separate commit in case this needs to be reverted in the future.

> As for Xtensa: Section 2.17.1 "Data Types and Alignment" of Xtensa LX Microprocessor Overview handbook https://loboris.eu/ESP32/Xtensa_lx%20Overview%20handbook.pdf says "`char` type is unsigned by default".
> 
> However, it appears that Xtensa has multiple ABIs. (Section 1.4 "Calling convention" of Overview of Xtensa ISA https://dl.espressif.com/github_assets/espressif/xtensa-isa-doc/releases/download/latest/Xtensa.pdf) (I have not checked whether the defaults are consistent among those ABIs or not -- this is why my PR has done nothing about Xtensa.)

Regarding Xtensa, my Clang change to make it unsigned by default was approved by Xtensa developers and has been merged now: https://github.com/llvm/llvm-project/pull/115967.

arichardson commented:
@rustbot review

tgross35 commented:
@rustbot label +I-libs-api-nominated

As noted in https://github.com/rust-lang/rust/issues/129945 there are a lot of tier 2 and 3 targets that disagree with Clang on the signedness of `c_char`, in particular the `-unknown-` targets. This PR changes Rust's logic to be more similar to Clang's and resolves most (all?) of the discrepancies. This is a correctness fix but it is breaking for the targets listed in the issue, so libs-api feedback is requested.

tgross35 commented:
Review notes: it is probably ideal to wait until https://github.com/rust-lang/rust/pull/133944 if that is ready in the near future, since it should give us a concrete list of what changed as part of this diff. Alternatively rerun @taiki-e's script from https://github.com/rust-lang/rust/issues/129945 before and after this change.

Could you pick up the long sources comment from https://github.com/rust-lang/rust/pull/131319/files#diff-92ac4f00ad02a9a9d7ebcd5337dec510097384904750feee32110bd614500050R95? Having somewhere specific to point to is probably good to preserve.

arichardson commented:
> Review notes: it is probably ideal to wait until #133944 if that is ready in the near future, since it should give us a concrete list of what changed as part of this diff. Alternatively rerun @taiki-e's script from #129945 before and after this change.
> 
> Could you pick up the long sources comment from https://github.com/rust-lang/rust/pull/131319/files#diff-92ac4f00ad02a9a9d7ebcd5337dec510097384904750feee32110bd614500050R95? Having somewhere specific to point to is probably good to preserve.

Sure will update this and and the appropriate attribution to the commit message. I ran the script before and after but didn't save the output. Will rerun it and paste the output compared to LLVM HEAD.

arichardson committed:
Add references to the specific ABI documents

Expcept for L4RE and Xtensa these were obtained from #131319

I could not find an open link to the Xtensa documentation, but the
signedness was confirmed by on of the Xtensa developers in
https://github.com/llvm/llvm-project/pull/115967#issuecomment-2506292323

Co-authored-by: Taiki Endo <te316e89@gmail.com>

Amanieu commented:
We discussed this in the libs-api meeting and we are OK with the breaking change since:
- it is actually a correctness fix.
- it primarily affects embedded targets (-unknown-) which are less likely to interact with C code.

arichardson commented:
I just finished running a slightly modified version of the test script from https://github.com/rust-lang/rust/issues/129945 

Summary:
- Fixes various aarch64/arm and powerpc results
  - I did have to fix a few incorrect PPC triples, will upload those fixes in a separate PR
- C codegen for XTensa is not enabled in upstream Clang yet (only LLVM and some parts of Clang have landed), but the rust value is correct.
- A few triples are not recognized by Clang: armv5te-unknown-linux-uclibcgnueabi, mips-unknown-linux-uclibc, mipsel-unknown-linux-uclibc, x86_64-unknown-l4re-uclibc, but the rust value has been validated against the sepcs.

Full output:

```
aarch64-apple-darwin: i8 correct
aarch64-apple-ios: i8 correct
aarch64-apple-ios-macabi: i8 correct
aarch64-apple-ios-sim: i8 correct
aarch64-apple-tvos: i8 correct
aarch64-apple-tvos-sim: i8 correct
aarch64-apple-visionos: i8 correct
aarch64-apple-visionos-sim: i8 correct
aarch64-apple-watchos: i8 correct
aarch64-apple-watchos-sim: i8 correct
aarch64-kmc-solid_asp3: u8 correct
aarch64-linux-android: u8 correct
aarch64-nintendo-switch-freestanding: u8 correct
aarch64-pc-windows-gnullvm: i8 correct
aarch64-pc-windows-msvc: i8 correct
aarch64-unknown-freebsd: u8 correct
aarch64-unknown-fuchsia: u8 correct
aarch64-unknown-hermit: u8 correct
aarch64-unknown-illumos: u8 correct
aarch64-unknown-linux-gnu: u8 correct
aarch64-unknown-linux-gnu_ilp32: u8 correct
aarch64-unknown-linux-musl: u8 correct
aarch64-unknown-linux-ohos: u8 correct
aarch64-unknown-netbsd: u8 correct
aarch64-unknown-none: u8 correct
aarch64-unknown-none-softfloat: u8 correct
aarch64-unknown-nto-qnx700: u8 correct
aarch64-unknown-nto-qnx710: u8 correct
aarch64-unknown-openbsd: u8 correct
aarch64-unknown-redox: u8 correct
aarch64-unknown-teeos: u8 correct
aarch64-unknown-trusty: u8 correct
aarch64-unknown-uefi: should be i8
aarch64-uwp-windows-msvc: i8 correct
aarch64-wrs-vxworks: u8 correct
aarch64_be-unknown-linux-gnu: u8 correct
aarch64_be-unknown-linux-gnu_ilp32: u8 correct
aarch64_be-unknown-netbsd: u8 correct
arm-linux-androideabi: u8 correct
arm-unknown-linux-gnueabi: u8 correct
arm-unknown-linux-gnueabihf: u8 correct
arm-unknown-linux-musleabi: u8 correct
arm-unknown-linux-musleabihf: u8 correct
arm64_32-apple-watchos: i8 correct
arm64e-apple-darwin: i8 correct
arm64e-apple-ios: i8 correct
arm64e-apple-tvos: i8 correct
arm64ec-pc-windows-msvc: i8 correct
armeb-unknown-linux-gnueabi: u8 correct
armebv7r-none-eabi: u8 correct
armebv7r-none-eabihf: u8 correct
armv4t-none-eabi: u8 correct
armv4t-unknown-linux-gnueabi: u8 correct
armv5te-none-eabi: u8 correct
armv5te-unknown-linux-gnueabi: u8 correct
armv5te-unknown-linux-musleabi: u8 correct
clang: error: version 'uclibcgnueabi' in target triple 'armv5te-unknown-linux-uclibcgnueabi' is invalid
clang: warning: unknown platform, assuming -mfloat-abi=soft
clang: warning: unknown platform, assuming -mfloat-abi=soft
clang: warning: unknown platform, assuming -mfloat-abi=soft
Failed to obtain builtin defines for armv5te-unknown-linux-uclibcgnueabi
armv5te-unknown-linux-uclibceabi: Unknown c_char
armv6-unknown-freebsd: u8 correct
armv6-unknown-netbsd-eabihf: u8 correct
armv6k-nintendo-3ds: u8 correct
armv7-linux-androideabi: u8 correct
armv7-rtems-eabihf: u8 correct
armv7-sony-vita-newlibeabihf: u8 correct
armv7-unknown-freebsd: u8 correct
armv7-unknown-linux-gnueabi: u8 correct
armv7-unknown-linux-gnueabihf: u8 correct
armv7-unknown-linux-musleabi: u8 correct
armv7-unknown-linux-musleabihf: u8 correct
armv7-unknown-linux-ohos: u8 correct
armv7-unknown-linux-uclibceabi: u8 correct
armv7-unknown-linux-uclibceabihf: u8 correct
armv7-unknown-netbsd-eabihf: u8 correct
armv7-unknown-trusty: u8 correct
armv7-wrs-vxworks-eabihf: u8 correct
armv7a-kmc-solid_asp3-eabi: u8 correct
armv7a-kmc-solid_asp3-eabihf: u8 correct
armv7a-none-eabi: u8 correct
armv7a-none-eabihf: u8 correct
armv7k-apple-watchos: i8 correct
armv7r-none-eabi: u8 correct
armv7r-none-eabihf: u8 correct
armv7s-apple-ios: i8 correct
armv8r-none-eabihf: u8 correct
clang: warning: no target microcontroller specified, please pass -mmcu=<mcu name> [-Wavr-rtlib-linking-quirks]
avr-unknown-gnu-atmega328: i8 correct
bpfeb-unknown-none: i8 correct
bpfel-unknown-none: i8 correct
clang: error: version 'abiv2' in target triple 'csky-unknown-linux-gnuabiv2' is invalid
Failed to obtain builtin defines for csky-unknown-linux-gnuabiv2
csky-unknown-linux-gnuabiv2: Unknown c_char
clang: error: version 'abiv2' in target triple 'csky-unknown-linux-gnuabiv2' is invalid
Failed to obtain builtin defines for csky-unknown-linux-gnuabiv2
csky-unknown-linux-gnuabiv2hf: Unknown c_char
hexagon-unknown-linux-musl: u8 correct
hexagon-unknown-none-elf: u8 correct
i386-apple-ios: i8 correct
i586-pc-nto-qnx700: i8 correct
i586-pc-windows-msvc: i8 correct
i586-unknown-linux-gnu: i8 correct
i586-unknown-linux-musl: i8 correct
i586-unknown-netbsd: i8 correct
i686-apple-darwin: i8 correct
i686-linux-android: i8 correct
i686-pc-windows-gnu: i8 correct
i686-pc-windows-gnullvm: i8 correct
i686-pc-windows-msvc: i8 correct
i686-unknown-freebsd: i8 correct
i686-unknown-haiku: i8 correct
i686-unknown-hurd-gnu: i8 correct
i686-unknown-linux-gnu: i8 correct
i686-unknown-linux-musl: i8 correct
i686-unknown-netbsd: i8 correct
i686-unknown-openbsd: i8 correct
i686-unknown-redox: i8 correct
i686-unknown-uefi: i8 correct
i686-uwp-windows-gnu: i8 correct
i686-uwp-windows-msvc: i8 correct
i686-win7-windows-msvc: i8 correct
i686-wrs-vxworks: i8 correct
loongarch64-unknown-linux-gnu: i8 correct
loongarch64-unknown-linux-musl: i8 correct
loongarch64-unknown-linux-ohos: i8 correct
loongarch64-unknown-none: i8 correct
loongarch64-unknown-none-softfloat: i8 correct
m68k-unknown-linux-gnu: i8 correct
mips-unknown-linux-gnu: i8 correct
mips-unknown-linux-musl: i8 correct
clang: error: version 'uclibc' in target triple 'mips-unknown-linux-uclibc' is invalid
Failed to obtain builtin defines for mips-unknown-linux-uclibc
mips-unknown-linux-uclibc: Unknown c_char
mips64-openwrt-linux-musl: i8 correct
mips64-unknown-linux-gnuabi64: i8 correct
mips64-unknown-linux-muslabi64: i8 correct
mips64el-unknown-linux-gnuabi64: i8 correct
mips64el-unknown-linux-muslabi64: i8 correct
mipsel-sony-psp: i8 correct
mipsel-sony-psx: i8 correct
mipsel-unknown-linux-gnu: i8 correct
mipsel-unknown-linux-musl: i8 correct
clang: error: version 'uclibc' in target triple 'mipsel-unknown-linux-uclibc' is invalid
Failed to obtain builtin defines for mipsel-unknown-linux-uclibc
mipsel-unknown-linux-uclibc: Unknown c_char
mipsel-unknown-netbsd: i8 correct
mipsel-unknown-none: i8 correct
mipsisa32r6-unknown-linux-gnu: i8 correct
mipsisa32r6el-unknown-linux-gnu: i8 correct
mipsisa64r6-unknown-linux-gnuabi64: i8 correct
mipsisa64r6el-unknown-linux-gnuabi64: i8 correct
msp430-none-elf: u8 correct
nvptx64-nvidia-cuda: i8 correct
powerpc-unknown-freebsd: u8 correct
powerpc-unknown-linux-gnu: u8 correct
powerpc-unknown-linux-gnuspe: u8 correct
powerpc-unknown-linux-musl: u8 correct
powerpc-unknown-linux-muslspe: u8 correct
powerpc-unknown-netbsd: u8 correct
powerpc-unknown-openbsd: u8 correct
powerpc-wrs-vxworks: u8 correct
powerpc-wrs-vxworks-spe: u8 correct
powerpc64-ibm-aix: u8 correct
powerpc64-unknown-freebsd: u8 correct
powerpc64-unknown-linux-gnu: u8 correct
powerpc64-unknown-linux-musl: u8 correct
powerpc64-unknown-openbsd: u8 correct
powerpc64-wrs-vxworks: u8 correct
powerpc64le-unknown-freebsd: u8 correct
powerpc64le-unknown-linux-gnu: u8 correct
powerpc64le-unknown-linux-musl: u8 correct
riscv32-wrs-vxworks: u8 correct
riscv32e-unknown-none-elf: u8 correct
riscv32em-unknown-none-elf: u8 correct
riscv32emc-unknown-none-elf: u8 correct
riscv32gc-unknown-linux-gnu: u8 correct
riscv32gc-unknown-linux-musl: u8 correct
riscv32i-unknown-none-elf: u8 correct
riscv32im-risc0-zkvm-elf: u8 correct
riscv32im-unknown-none-elf: u8 correct
riscv32ima-unknown-none-elf: u8 correct
riscv32imac-esp-espidf: u8 correct
riscv32imac-unknown-none-elf: u8 correct
riscv32imac-unknown-nuttx-elf: u8 correct
riscv32imac-unknown-xous-elf: u8 correct
riscv32imafc-esp-espidf: u8 correct
riscv32imafc-unknown-none-elf: u8 correct
riscv32imafc-unknown-nuttx-elf: u8 correct
riscv32imc-esp-espidf: u8 correct
riscv32imc-unknown-none-elf: u8 correct
riscv32imc-unknown-nuttx-elf: u8 correct
riscv64-linux-android: u8 correct
riscv64-wrs-vxworks: u8 correct
riscv64gc-unknown-freebsd: u8 correct
riscv64gc-unknown-fuchsia: u8 correct
riscv64gc-unknown-hermit: u8 correct
riscv64gc-unknown-linux-gnu: u8 correct
riscv64gc-unknown-linux-musl: u8 correct
riscv64gc-unknown-netbsd: u8 correct
riscv64gc-unknown-none-elf: u8 correct
riscv64gc-unknown-nuttx-elf: u8 correct
riscv64gc-unknown-openbsd: u8 correct
riscv64imac-unknown-none-elf: u8 correct
riscv64imac-unknown-nuttx-elf: u8 correct
s390x-unknown-linux-gnu: u8 correct
s390x-unknown-linux-musl: u8 correct
sparc-unknown-linux-gnu: i8 correct
sparc-unknown-none-elf: i8 correct
sparc64-unknown-linux-gnu: i8 correct
sparc64-unknown-netbsd: i8 correct
sparc64-unknown-openbsd: i8 correct
sparcv9-sun-solaris: i8 correct
thumbv4t-none-eabi: u8 correct
thumbv5te-none-eabi: u8 correct
thumbv6m-none-eabi: u8 correct
thumbv6m-nuttx-eabi: u8 correct
thumbv7a-pc-windows-msvc: i8 correct
thumbv7a-uwp-windows-msvc: i8 correct
thumbv7em-none-eabi: u8 correct
thumbv7em-none-eabihf: u8 correct
thumbv7em-nuttx-eabi: u8 correct
thumbv7em-nuttx-eabihf: u8 correct
thumbv7m-none-eabi: u8 correct
thumbv7m-nuttx-eabi: u8 correct
thumbv7neon-linux-androideabi: u8 correct
thumbv7neon-unknown-linux-gnueabihf: u8 correct
thumbv7neon-unknown-linux-musleabihf: u8 correct
thumbv8m.base-none-eabi: u8 correct
thumbv8m.base-nuttx-eabi: u8 correct
thumbv8m.main-none-eabi: u8 correct
thumbv8m.main-none-eabihf: u8 correct
thumbv8m.main-nuttx-eabi: u8 correct
thumbv8m.main-nuttx-eabihf: u8 correct
wasm32-unknown-emscripten: i8 correct
wasm32-unknown-unknown: i8 correct
wasm32-wasip1: i8 correct
wasm32-wasip1-threads: i8 correct
wasm32-wasip2: i8 correct
wasm32v1-none: i8 correct
wasm64-unknown-unknown: i8 correct
x86_64-apple-darwin: i8 correct
x86_64-apple-ios: i8 correct
x86_64-apple-ios-macabi: i8 correct
x86_64-apple-tvos: i8 correct
x86_64-apple-watchos-sim: i8 correct
x86_64-fortanix-unknown-sgx: i8 correct
x86_64-linux-android: i8 correct
x86_64-pc-nto-qnx710: i8 correct
x86_64-pc-solaris: i8 correct
x86_64-pc-windows-gnu: i8 correct
x86_64-pc-windows-gnullvm: i8 correct
x86_64-pc-windows-msvc: i8 correct
x86_64-unikraft-linux-musl: i8 correct
x86_64-unknown-dragonfly: i8 correct
x86_64-unknown-freebsd: i8 correct
x86_64-unknown-fuchsia: i8 correct
x86_64-unknown-haiku: i8 correct
x86_64-unknown-hermit: i8 correct
x86_64-unknown-hurd-gnu: i8 correct
x86_64-unknown-illumos: i8 correct
clang: error: version 'uclibc' in target triple 'x86_64-unknown-l4re-uclibc' is invalid
Failed to obtain builtin defines for x86_64-unknown-l4re-uclibc
x86_64-unknown-l4re-uclibc: Unknown c_char
x86_64-unknown-linux-gnu: i8 correct
x86_64-unknown-linux-gnux32: i8 correct
x86_64-unknown-linux-musl: i8 correct
x86_64-unknown-linux-none: i8 correct
x86_64-unknown-linux-ohos: i8 correct
x86_64-unknown-netbsd: i8 correct
x86_64-unknown-none: i8 correct
x86_64-unknown-openbsd: i8 correct
x86_64-unknown-redox: i8 correct
x86_64-unknown-trusty: i8 correct
x86_64-unknown-uefi: i8 correct
x86_64-uwp-windows-gnu: i8 correct
x86_64-uwp-windows-msvc: i8 correct
x86_64-win7-windows-msvc: i8 correct
x86_64-wrs-vxworks: i8 correct
x86_64h-apple-darwin: i8 correct
error: unknown target triple 'xtensa-unknown-none-elf'
Failed to obtain builtin defines for xtensa-none-elf
xtensa-esp32-espidf: Unknown c_char
error: unknown target triple 'xtensa-unknown-none-elf'
Failed to obtain builtin defines for xtensa-none-elf
xtensa-esp32-none-elf: Unknown c_char
error: unknown target triple 'xtensa-unknown-none-elf'
Failed to obtain builtin defines for xtensa-none-elf
xtensa-esp32s2-espidf: Unknown c_char
error: unknown target triple 'xtensa-unknown-none-elf'
Failed to obtain builtin defines for xtensa-none-elf
xtensa-esp32s2-none-elf: Unknown c_char
error: unknown target triple 'xtensa-unknown-none-elf'
Failed to obtain builtin defines for xtensa-none-elf
xtensa-esp32s3-espidf: Unknown c_char
error: unknown target triple 'xtensa-unknown-none-elf'
Failed to obtain builtin defines for xtensa-none-elf
xtensa-esp32s3-none-elf: Unknown c_char
```

tgross35 commented:
It looks like `aarch64-unknown-uefi: should be i8` is the last one that still has a mismatch, do you know why that would be?

tgross35 commented:
Pinging some target maintainers:

- Solid @kawadakk
- Hermit @stlankes @mkroening
- Illumos @jclulow @sunshowers 
- Redox @jackpot51 
- Trusty @randomPoison @ahomescu 
- Hexagon @androm3da
- RISC-V @robin-randhawa-sifive @raw-bin
- RISC-V Android @chriswalles @maurer 
- Xous @xobs
- ESP-IDF @ivmarkov @MabezDev
- Nuttx @no1wudi 

If you were pinged here, this PR will change one of your target's definition of `c_char` to be in line with Clang's, most from `i8` to `u8`. Not all targets change, this is pretty limited to Arm and RISC-V. The full list of changing targets is at https://github.com/rust-lang/rust/issues/129945 and the list after being fixed is https://github.com/rust-lang/rust/pull/132975#issuecomment-2532688643, let us know if you know something to be incorrect here.

jclulow commented:
>     * it primarily affects embedded targets (-unknown-) which are less likely to interact with C code.

I don't think that's actually what **-unknown-** means, FWIW.  Looking at (the platform support list)[https://doc.rust-lang.org/nightly/rustc/platform-support.html] I see quite a lot of targets that contain the string **-unknown-** and are not embedded systems.

jclulow commented:
I have confirmed with the folks working on the illumos ARM port that **char** is unsigned there to match the prevailing winds, so this makes sense to us, thanks!

randomPoison commented:
This is correct for Trusty as far as I know, we expect `c_char` to be `u8` on both `aarch64` and `arm`.

arichardson commented:
> aarch64-unknown-uefi

Interesting, it looks like this is probably a bug in the target definition, which means it uses the windows ABI in Clang:
`        llvm_target: "aarch64-unknown-windows".into(),`

This should be "aarch64-unknown-uefi" which Clang understands
```
❯ ~/upstream-llvm-project/cmake-build-debug/bin/clang -E -dM -x c /dev/null -target aarch64-unknown-uefi | grep __CHAR_UN
#define __CHAR_UNSIGNED__ 1
❯ ~/upstream-llvm-project/cmake-build-debug/bin/clang -E -dM -x c /dev/null -target aarch64-unknown-windows | grep __CHAR_UN
```

jackpot51 commented:
This is fine for Redox OS.

tgross35 commented:
> > aarch64-unknown-uefi
> 
> Interesting, it looks like this is probably a bug in the target definition, which means it uses the windows ABI in Clang: ` llvm_target: "aarch64-unknown-windows".into(),`
> 
> This should be "aarch64-unknown-uefi" which Clang understands

That is mildly amusing. Opened https://github.com/rust-lang/rust/pull/134156 to fix it.

tgross35 commented:
With the libs-api approval I don't think there is any reason to hold off, especially considering it was mentioned during the meeting that this is better to have toward the beginning of a release cycle (we are reasonably early now, next branch is January 3rd for stable on February 20th).

@bors r+

Thanks everyone who took a look at this.

bors commented:
:pushpin: Commit dd3e98c58bc94533e875ac1d2b87163727d59cd2 has been approved by `tgross35`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=tgross35 dd3e98c58bc94533e875ac1d2b87163727d59cd2 -->
<!-- homu: {"type":"Approved","sha":"dd3e98c58bc94533e875ac1d2b87163727d59cd2","approver":"tgross35","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

arichardson commented:
> > > aarch64-unknown-uefi
> > 
> > 
> > Interesting, it looks like this is probably a bug in the target definition, which means it uses the windows ABI in Clang: ` llvm_target: "aarch64-unknown-windows".into(),`
> > This should be "aarch64-unknown-uefi" which Clang understands
> 
> That is mildly amusing. Opened #134156 to fix it.

Digging deeper it looks like this is only partially supported in Clang (x86 supports the triple, aarch64 triggers assertions). I do wonder if Clang's output here is wrong and Rust is correct - I imagine UEFI follows the Windows ABIs since it is COFF-based? But I haven't found a spec for the UEFI AArch64 ABI.

androm3da commented:
Confirmed: for `hexagon` targets, `c_char` should be unsigned. ✅

Thanks for the fix.

ivmarkov commented:
> Pinging some target maintainers (checkbox = acked):
> 
> * [ ]  Solid @kawadakk
> * [ ]  Hermit @stlankes @mkroening
> * [x]  Illumos @jclulow @sunshowers
> * [x]  Redox @jackpot51
> * [x]  Trusty @randomPoison @ahomescu
> * [x]  Hexagon @androm3da
> * [ ]  RISC-V @robin-randhawa-sifive @raw-bin
> * [ ]  RISC-V Android @chriswalles @maurer
> * [x]  Xous @xobs
> * [ ]  ESP-IDF @ivmarkov @MabezDev
> * [ ]  Nuttx @no1wudi
> 
> If you were pinged here, this PR will change one of your target's definition of `c_char` to be in line with Clang's, most from `i8` to `u8`. Not all targets change, this is pretty limited to Arm and RISC-V. The full list of changing targets is at #129945 and the list after being fixed is [#132975 (comment)](https://github.com/rust-lang/rust/pull/132975#issuecomment-2532688643), let us know if you know something to be incorrect here.

It looks like indeed `c_char` should be unsigned for both xtensa and riscv ESP-IDF targets. Yet - we are likely to have downstream breakages because of this change, as I noticed there are (a few) places where we have frivolously used `c_char` and `i8` interchangeably, and obviously these places need to be fixed to only use `core::ffi::c_char` now.

With that said, I guess it is on us to fix this downstream, and we'll know the impact pretty soon, given the "bors +r" from a couple of hours ago.

mkroening commented:
This aligns Hermit to its [definitions in `libc`](https://github.com/rust-lang/libc/blob/0.2.168/src/hermit.rs#L5-L11), so this is good from our side. Thanks! :)

ivmarkov commented:
Hmm, looks like we need to fix then the ESP IDF [`c_char` definition in `libc`](https://github.com/rust-lang/libc/blob/e6082ef26096154e08ab0cb094d66d5ca116ee15/src/unix/newlib/espidf/mod.rs#L4), as it is `i8` and this would now be different from the `core::ffi::c_char` definition.

I also wonder (excuse my ignorance) what this change means with regards to interop with C code compiled with GCC, [as it seems GCC always treats `char` as a signed char, unless you tell it otherwise](https://stackoverflow.com/a/2054959)?

tgross35 commented:
> Hmm, looks like we need to fix then the ESP IDF [`c_char` definition in `libc`](https://github.com/rust-lang/libc/blob/e6082ef26096154e08ab0cb094d66d5ca116ee15/src/unix/newlib/espidf/mod.rs#L4), as it is `i8` and this would now be different from the `core::ffi::c_char` definition.

`libc` will need to be updated after this, most of the targets listed here probably have the same definition in `libc`.

> I also wonder (excuse my ignorance) what this change means with regards to interop with C code compiled with GCC, [as it seems GCC always treats `char` as a signed char, unless you tell it otherwise](https://stackoverflow.com/a/2054959)?

I think this answer might just be referring to x86, https://gcc.gnu.org/onlinedocs/gcc-4.2.2/gcc/C-Dialect-Options.html says:

> Each kind of machine has a default for what char should be. It is either like unsigned char by default or like signed char by default. 

And based on testing a few targets at https://godbolt.org/z/vxvY34W8b, at least the ARM and RISC-V targets seem to have __CHAR_UNSIGNED__ defined.

AVR GCC actually does not have this defined for the thumb targets so we might need to double check those https://godbolt.org/z/rdroc9fe9 (@Patryk27).

> It looks like indeed `c_char` should be unsigned for both xtensa and riscv ESP-IDF targets. Yet - we are likely to have downstream breakages because of this change, as I noticed there are (a few) places where we have frivolously used `c_char` and `i8` interchangeably, and obviously these places need to be fixed to only use `core::ffi::c_char` now.
> 
> With that said, I guess it is on us to fix this downstream, and we'll know the impact pretty soon, given the "bors +r" from a couple of hours ago.

To be clear, we can make followup changes to specific targets as needed, or if the fallout is especially bad in particular areas.

taiki-e commented:
@tgross35 

> AVR GCC actually does not have this defined for the thumb targets so we might need to double check those

thumb is thumb mode of 32-bit Arm (`target_arch = "arm"`). AVR GCC is for AVR (`target_arch = "avr"`) and unrelated to thumb.

c_char of AVR is signed by default as mentioned in the references in https://github.com/rust-lang/rust/issues/129945.

>AVR
https://gcc.gnu.org/wiki/avr-gcc#Type_Layout says char is signed

kawadakk commented:
The SOLID toolchain actually uses `-fsigned-char`, but we are planning to phase out this flag in a near feature, and the impact is limited due to the nature of embedded systems, so we are okay with this change.

ids1024 commented:
This also seems to break compilation for nuttx targets, since it doesn't match `libc`, but it probably is right and should be fixed there.

bugadani commented on library/core/src/ffi/mod.rs:
```diff
@@ -91,59 +91,86 @@ pub type c_ssize_t = isize;
+        //   - z/OS: XL C/C++ Language Reference says: "By default, char behaves like an unsigned char."
+        //     https://www.ibm.com/docs/en/zos/3.1.0?topic=specifiers-character-types
+        // Xtensa:
+        //   - "The char type is unsigned by default for Xtensa processors."
```
Where is this information coming from?

ivmarkov commented:
> To be clear, we can make followup changes to specific targets as needed, or if the fallout is especially bad in particular areas.

No need to. 
I've now checked the `ESP-IDF` checkbox in that `core::ffi::c_char` (and `libc`'s `c_char`) should indeed both be unsigned, as  per [these findings](https://github.com/rust-lang/libc/pull/4195#issuecomment-2543096509).

nico commented:
This changed c_char on android/riscv from signed to unsinged. Was that intentional?

If so, https://github.com/rust-lang/libc/blob/main/src/unix/linux_like/android/b64/riscv64/mod.rs probably needs updating?

arichardson commented:
> This changed c_char on android/riscv from signed to unsinged. Was that intentional?
> 
> If so, https://github.com/rust-lang/libc/blob/main/src/unix/linux_like/android/b64/riscv64/mod.rs probably needs updating?

Not explicitly intentional, but a correct change. Prior to this rust disagreed with clang, now it matches (see https://github.com/rust-lang/rust/pull/132975#issuecomment-2532688643)

nico commented:
Thoughts on the second part of the question? c_char in https://github.com/rust-lang/libc/blob/main/src/unix/linux_like/android/b64/riscv64/mod.rs now doesn't match c_char here, which leads to build errors on android/riscv64 (e.g. https://crbug.com/384413087)

arichardson commented:
> Thoughts on the second part of the question? c_char in https://github.com/rust-lang/libc/blob/main/src/unix/linux_like/android/b64/riscv64/mod.rs now doesn't match c_char here, which leads to build errors on android/riscv64 (e.g. https://crbug.com/384413087)

Yes, it looks like the other c_char needs to be updated to u8. Sorry about all the breakage this pr is causing...

tgross35 commented:
> Thoughts on the second part of the question? c_char in https://github.com/rust-lang/libc/blob/main/src/unix/linux_like/android/b64/riscv64/mod.rs now doesn't match c_char here, which leads to build errors on android/riscv64 (e.g. https://crbug.com/384413087)

Anything that changed here does need to be updated at `libc`. I'll try to make sure these are fixed and in a release next week.

As a temporary workaround, can your build patch an `as *const libc::c_char` in `stdarch` so it keeps building regardless of whether or not `libc` has been updated? I think `riscv64-linux-android` is the only affected target that needs the `linux` modules in `stdarch`.

(sorry, had a comment typed but forgot to submit)

tgross35 commented:
Updates to `libc` were released in https://github.com/rust-lang/libc/releases/tag/0.2.169.
